### PR TITLE
Constrain base.in versions with edx-platform pins

### PR DESCRIPTION
**Pull request by nedbat**
**Description:**

To ensure that this repo is tested with the same versions of libraries it will eventually run with, `make upgrade` now pulls down an edx-platform requirements file and uses it to constrain versions here.

I've removed the `test-master.in` file, which was meant to serve the same purpose. But it had to be manually updated, and so didn't match in a few places.

I've removed the `test-hawthorn` files, since Hawthorn is unsupported, and we don't have Ironwood files.

**JIRA:** None

**Dependencies:** None

**Merge deadline:** Soon?

**Installation instructions:** Nothing unusual

**Testing instructions:**

1. Open page A
2. Do thing B
3. Expect C to happen
4. If D happend instead - check failed.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] Called `make static` for webpack bundling if any static content was updated.
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

### CODE REVIEW

File: requirements/base.in [hunk 1]
```diff
@@ -1,12 +1,46 @@
...
+git+https://github.com/edx/django-rest-framework-oauth.git@0a43e8525f1e3048efe4bc70c03de308a277197c#egg=djangorestframework-oauth==1.1.1
+djangorestframework-xml
+django-multi-email-field            # Multi email input field in Django admin
+jsondiff                            # Used to detect content metadata changes
+django-countries
 -e git+https://github.com/edx/TinCanPython.git@0.0.5.py.35#egg=tincan==0.0.5.py.35
-django-simple-history==2.7.0
-edx-rbac==0.2.1
+django-simple-history
+edx-rbac
```
Comment (christopappas) [inline requirements/base.in]: @brittneyexline is this safe to unpin? i've lost track of that conversation

Comment (christopappas) [reply requirements/base.in]: (and `edx-drf-extensions`)?

File: Makefile [hunk 2]
```diff
@@ -76,19 +73,20 @@ docs: ## generate Sphinx HTML documentation, including API docs
 
 # Define PIP_COMPILE_OPTS=-v to get more information during make upgrade.
 PIP_COMPILE = pip-compile --upgrade $(PIP_COMPILE_OPTS)
```
Comment (jmbowman) [inline Makefile]: We've started also using the `--rebuild` option; from the latest revision to OEP-18:

> By default pip-compile uses a cache of calculated dependency relationships to improve the performance of subsequent runs. Unfortunately, the results of this cache are sometimes used even after a new package release has changed the set of packages it depends on. To avoid generating incorrect requirements files due to this, it’s best to always use the --rebuild option when running pip-compile.

File: Makefile [hunk 3]
```diff
@@ -76,19 +73,20 @@ docs: ## generate Sphinx HTML documentation, including API docs
...
+CONSTRAINTS = requirements/edx-platform-constraints.txt
 
-upgrade: ## update the requirements/*.txt files with the latest packages satisfying requirements/*.in
+upgrade: export CUSTOM_COMPILE_COMMAND=make upgrade
+upgrade:  ## update the requirements/*.txt files with the latest packages satisfying requirements/*.in
+	echo "### DON'T edit this file, it's copied from edx-platform. See make upgrade" > $(CONSTRAINTS)
+	curl -fsSL https://raw.githubusercontent.com/edx/edx-platform/master/requirements/edx/base.txt | grep -v '^-e' >> $(CONSTRAINTS)
```
Comment (christopappas) [inline Makefile]: 👍

File: Makefile [hunk 4]
```diff
@@ -76,19 +73,20 @@ docs: ## generate Sphinx HTML documentation, including API docs
...
 	pip install -q pip-tools
-	$(PIP_COMPILE) -o requirements/dev.txt requirements/base.in requirements/test-$(TARGET_PLATFORM).in requirements/dev.in requirements/quality.in
-	$(PIP_COMPILE) -o requirements/doc.txt requirements/base.in requirements/test-$(TARGET_PLATFORM).in requirements/doc.in
-	$(PIP_COMPILE) -o requirements/quality.txt requirements/base.in requirements/test-$(TARGET_PLATFORM).in requirements/dev.in requirements/quality.in requirements/doc.in requirements/test.in
+	$(PIP_COMPILE) -o requirements/base.txt requirements/base.in
+	$(PIP_COMPILE) -o requirements/dev.txt requirements/dev.in requirements/quality.in
```
Comment (jmbowman) [inline Makefile]: Do we still need a distinction between `dev.txt` and `quality.txt` here?  It looks like the former is a subset of the latter, and I'd think that developers would want to be able to run any of the quality checks in their dev environment.

File: Makefile [hunk 5]
```diff
@@ -76,19 +73,20 @@ docs: ## generate Sphinx HTML documentation, including API docs
...
+	$(PIP_COMPILE) -o requirements/doc.txt requirements/doc.in
+	$(PIP_COMPILE) -o requirements/quality.txt requirements/dev.in requirements/quality.in requirements/doc.in requirements/test.in
```
Comment (jmbowman) [inline Makefile]: Could `quality.in` be updated to include `doc.txt`, etc. instead?  We've found that listing multiple `*.in` files for the same output file can lead to different versions of the same package being listed in the different output files (which isn't good for consistency between production, testing, and development).

File: package-lock.json [hunk 6]
```diff
@@ -1794,7 +1794,8 @@
         "ansi-regex": {
           "version": "2.1.1",
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
```
Comment (jmbowman) [inline package-lock.json]: I don't see a corresponding change in `package.json`; were these changes intentional, or an accident of clashing npm versions?

**Comment by christopappas** [comment]
Follow up question for you @nedbat 

The case where we decide to upgrade a package in edx-enterprise that also is in the constraints file we create from edx-platform.

Say, currently both edx-enterprise and edx-platform have edx-rbac==0.2.1.  If we decide to update edx-rbac to version 1.0.X, previously we would have run make upgrade for edx-enterprise which bumps a number of version of packages, create a new egg/release, and then run make upgrade in edx-platform, which would in turn update a number of the same repos.

The question: do we now need to run make upgrade in edx-platform and have a PR merged first to bump the versions of packages that edx-enterprise uses?

If we don't, when we try to run make upgrade in edx-enterprise, we are going to be constrained on old versions of those packages, right? So even if we wanted to get edx-rbac 1.0.X in edx-enterprise, edx-platform still will have `edx_rbac==0.2.1` as a constraint

Does this sound right?

**Comment by nedbat** [comment]
@christopappas Thanks for noticing the edx-rbac issue.  I think this is a fatal flaw, that requirements flow from edx-enterprise to edx-platform, and then back.  There are some fancier ideas of what to do about that, but I think I will change this to a much simpler system that simply warns if the pins here and in edx-platform are out of sync.

File: requirements/base.in (hunk 1 above)
Comment (brittneyexline) [reply requirements/base.in]: we cannot unpin edx-rbac until ENT-2002 is addressed, so we should keep this at `0.2.1`

**Comment by nedbat** [comment]
Closed in favor of #507
